File: bench/node/errors-deep.js
const n = parseInt(process.env.BENCH_N || "50000", 10);
const md = 1000000007;
const DEPTH = 50;

class BenchError extends Error {
  constructor(v) { super(); this.v = v; }
}

function descend(d, i) {
  if (d === 0) throw new BenchError(i % 100);
  return 1 + descend(d - 1, i);
}

let acc = 0;
for (let i = 0; i < n; i++) {
  try { descend(DEPTH, i); }
  catch (e) { acc += e.v; }
}
console.log(acc % md);

File: bench/erlang/errors-deep.erl
#!/usr/bin/env escript
%%! -mode(compile)
%% error propagation: throw `depth` non-tail frames down, catch at the top, N times.
main(_) ->
    N = bench_n(50000), Md = 1000000007, Depth = 50,
    Acc = lists:foldl(fun(I, Acc) ->
        try descend(Depth, I)
        catch throw:{bench_error, V} -> Acc + V end
    end, 0, lists:seq(0, N-1)),
    io:format("~w~n", [Acc rem Md]).
bench_n(D) -> case os:getenv("BENCH_N") of false -> D; V -> list_to_integer(V) end.
descend(0, I) -> throw({bench_error, I rem 100});
descend(D, I) -> 1 + descend(D-1, I).
